server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java

```diff
@@ -319,21 +324,30 @@ private static void updateRootTabletDataFile(KeyExtent extent, FileRef path, Fil
    *
    * @return A Mutation to update a tablet from the given information
    */
-  private static Mutation getUpdateForTabletDataFile(KeyExtent extent, FileRef path,
-      FileRef mergeFile, DataFileValue dfv, String time, Set<FileRef> filesInUseByScans,
-      String address, ZooLock zooLock, Set<String> unusedWalLogs, TServerInstance lastLocation,
-      long flushId) {
+  private static Mutation getUpdateForTabletDataFile(ClientContext context, KeyExtent extent,
+      FileRef path, FileRef mergeFile, DataFileValue dfv, String time,
+      Set<FileRef> filesInUseByScans, String address, ZooLock zooLock, Set<String> unusedWalLogs,
+      TServerInstance lastLocation, long flushId) {
     Mutation m = new Mutation(extent.getMetadataEntry());
 
     if (dfv.getNumEntries() > 0) {
       m.put(DataFileColumnFamily.NAME, path.meta(), new Value(dfv.encode()));
       TabletsSection.ServerColumnFamily.TIME_COLUMN.put(m, new Value(time.getBytes(UTF_8)));
-      // stuff in this location
-      TServerInstance self = getTServerInstance(address, zooLock);
-      self.putLastLocation(m);
-      // erase the old location
-      if (lastLocation != null && !lastLocation.equals(self))
-        lastLocation.clearLastLocation(m);
+
+      // if the location mode is 'locality'', then preserve the current compaction location in the
+      // last location value
+      if ("locality".equals(context.getConfiguration().get(Property.TABLE_LOCATION_MODE))) {
+        // stuff in this location
+        TServerInstance self = getTServerInstance(address, zooLock);
+        self.putLastLocation(m);
+        // erase the old location
+        if (lastLocation != null && !lastLocation.equals(self))
+          // @TODO: this does not make any sense to me as this will simply clear out the location we
+          // just set a couple lines previously
+          // clearLastLocation will clear the "last" location no matter what the value is
+          lastLocation.clearLastLocation(m);
```

Comment:
This is setting the last location on a completely different tserver. It's updating the mutation by putting the new tserver as the last location, and clearing the previous tserver as the last location. One is a put, and one is a delete, on two separate metadata columns. They have the same column family ("last"), but different column qualifiers (the tserver hostname). We check that they aren't the same, so we're not adding a `delete` to the same mutation that we just added a `put` for the same column.

Reply:
well crud.  I did not read the code correctly.  thanks.